pkg/network/node/iptables.go

```diff
@@ -167,23 +153,56 @@ func (n *NodeIPTables) getNodeIPTablesChains() []Chain {
 				{"-i", "docker0", "-m", "comment", "--comment", "from docker to localhost", "-j", "ACCEPT"},
 			},
 		},
-		{
-			table:    "filter",
-			name:     "OPENSHIFT-FIREWALL-FORWARD",
-			srcChain: "FORWARD",
-			srcRule:  []string{"-m", "comment", "--comment", "firewall overrides"},
-			rules: [][]string{
-				{"-s", n.clusterNetworkCIDR, "-m", "comment", "--comment", "attempted resend after connection close", "-m", "conntrack", "--ctstate", "INVALID", "-j", "DROP"},
-				{"-d", n.clusterNetworkCIDR, "-m", "comment", "--comment", "forward traffic from SDN", "-j", "ACCEPT"},
-				{"-s", n.clusterNetworkCIDR, "-m", "comment", "--comment", "forward traffic to SDN", "-j", "ACCEPT"},
-			},
-		},
-		{
+		Chain{
 			table:    "filter",
 			name:     "OPENSHIFT-ADMIN-OUTPUT-RULES",
 			srcChain: "FORWARD",
 			srcRule:  []string{"-i", Tun0, "!", "-o", Tun0, "-m", "comment", "--comment", "administrator overrides"},
 			rules:    nil,
-		},
+		})
+
+	var masqRules [][]string
+	var masq2Rules [][]string
+	var filterRules [][]string
+	for _, cidr := range n.clusterNetworkCIDR {
+		if n.masqueradeServices {
+			masqRules = append(masqRules, []string{"-s", cidr, "-m", "comment", "--comment", "masquerade pod-to-service and pod-to-external traffic", "-j", "MASQUERADE"})
+		} else {
+			masqRules = append(masqRules, []string{"-s", cidr, "-m", "comment", "--comment", "masquerade pod-to-external traffic", "-j", "OPENSHIFT-MASQUERADE-2"})
+			masq2Rules = append(masq2Rules, []string{"-d", cidr, "-m", "comment", "--comment", "masquerade pod-to-external traffic", "-j", "RETURN"})
+		}
+
+		filterRules = append(filterRules, []string{"-s", cidr, "-m", "comment", "--comment", "attempted resend after connection close", "-m", "conntrack", "--ctstate", "INVALID", "-j", "DROP"})
+		filterRules = append(filterRules, []string{"-d", cidr, "-m", "comment", "--comment", "forward traffic from SDN", "-j", "ACCEPT"})
+		filterRules = append(filterRules, []string{"-s", cidr, "-m", "comment", "--comment", "forward traffic to SDN", "-j", "ACCEPT"})
 	}
+
+	if !n.masqueradeServices {
+		masq2Rules = append(masq2Rules, []string{"-j", "MASQUERADE"})
+		chainArray = append(chainArray,
+			Chain{
+				table:    "nat",
+				name:     "OPENSHIFT-MASQUERADE-2",
+				srcChain: "POSTROUTING",
+				srcRule:  []string{"-m", "comment", "--comment", "additional rules for masquerading Openshift traffic"},
```

Comment:
This will create a rule in POSTROUTING jumping to OPENSHIFT-MASQUERADE-2. You want to leave srcChain and srcRule unset here (but you'll have to also update the code in `syncIPTablesRules` to handle that case, where it has to make sure the chain exists, but not create a rule jumping to it).